### DISCUSSION TITLE
[Feature Request] Pointed Bookmarks

### DISCUSSION THREAD

**Discussion by redwizard42** [RetroAchievements/RAIntegration#1098] [Ideas]
It would be very helpful to bookmark something and the end of a pointer chain for purposes of pausing on change or freezing the value at the current resolved address.

My initial though is that since the current bookmark format is just the logic string of the memory reference, that string could be more expansive logic: ``I:0xX123456&0x7fffffff_I:0xX00001c&0x7fffffff_0xH00002d`` for a chain. This isn't exactly valid logic currently since there's no measurement or right side on the last condition, so additional thoughts are welcome.

Would likely need to add a button to add such a bookmark, with an asset-editor-like interface to specify each part of the chain and any scaling (bitmasking or for indexing--multiplication, most likely).

My expectation would be that the address field would display the resulting address of the chain, pausing would check if the resolved value at the end of the chain is different from last frame, and freezing would freeze the current address's value if different from previous frame.

Thoughts on implementation welcome.

**Comment by Jamiras** [comment]
I was stalling on this waiting for variables.

A variable containing the pointer chain could be used in the achievement logic, complete with delta/prior support, and the variable could just be added to the bookmarks list.

Regardless of the implementation, the Pause on change functionality will be easy. Freezing the value is harder as the write-back code would have to traverse the pointer chain just like the read code in rcheevos does.

**Reply by redwizard42** [reply]
If implemented without variables, after the read the address of the value is known and could be stored and so the frozen value could be written directly to that.  The tradeoff of course is that the read has to traverse the chain.  A benefit of adding functionality without it being a bookmark of a variable is that you wouldn't have to define a variable (and any logic that sets it) just to bookmark something you may be interested in during development, but ultimately don't need to use for logic.
